src/librustc_trait_selection/traits/structural_match.rs

```diff
@@ -251,7 +251,10 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
         // fields of ADT.
         let tcx = self.tcx();
         for field_ty in adt_def.all_fields().map(|field| field.ty(tcx, substs)) {
-            if field_ty.visit_with(self) {
+            let ty = self.tcx().normalize_erasing_regions(ty::ParamEnv::empty(), field_ty);
```

Comment:
`ty::Param::empty()` is incorrect - frankly I wonder if we should remove it.
You should have a `ParamEnv` around.

(*couple minutes later*)

Hang on, this entire file is missing real `ParamEnv`s? That's liable to cause ICEs if generic parameters end up in these types.

Reply:
Does this block this PR? Afaik it's currently not possible on stable to hit this case.

Reply:
I may go ahead and fix this afterwards, but my current priority is mainly to fix #72896 
Which probably means that we have to backport this to beta.

Another option would be to revert the structural match check for projections :thinking:

Reply:
Its probably my fault that we don't have `ParamEnv`s threaded through here. This whole thing of switching from attributes to traits has been a real rabbit hole.

Still, I think landing this is better than reverting the structural match check for projections.